### PR TITLE
Application of Byzantine Consensus Mechanisms 

### PR DESCRIPTION
**Pull request by Kevoulee**
@CjS77 @hansieodendaal MainReport is ready for review.

### CODE REVIEW

**Comment by hansieodendaal** [comment]
My bad: "_Also no need to add anchors everywhere. This is done automatically and makes readability of `.md` file poor_" Could not get the TOC to work properly if special characters were included in the headings like '?', ',' or '&' so I introduced the anchors. I see now this is solved by just omitting all of those.

File: src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md [hunk 1]
```diff
@@ -0,0 +1,739 @@
...
+
+Now despite assuming a synchronous network; which is susceptible to the DoS attacks, as noted by Dr Emin Gün Sirer, the new family of protocols "reaches a level of security that is simply good enough while surging forward with other advancements"  
+
+### PARSEC
+
+PARSEC is a byzantine fault tolerant consensus algorithm possessing **weak synchrony** assumptions (highly asynchronous, assuming random delays with finite expected value)
+
+Similar to Hashgraph, it has no leaders, no round robin, no proof of work and reaches eventual consensus with probability one. If differs from Hashgraph, in that it provides high speed in the absence and presence of faults.  Thus, it avoids the structures of delegated proof of stake, which requires a trusted set of leaders, and does not have a round robin (where a set of instated miners sign each block)
```
Comment (CjS77) [inline src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md]: instated?

File: src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md [hunk 2]
```diff
@@ -0,0 +1,739 @@
...
+
+It is fully open, unlike Hashgraph, which is patented and closed sourced and the GPLv3 written in Rust will soon be available. [[1]]
```
Comment (CjS77) [inline src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md]: It is available.

-> The reference implementation of PARSEC, written in Rust, was released a few weeks after the whitepaper [[PSC]].

[PSC]: https://github.com/maidsafe/parsec 'PARSEC Github repository'

File: src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md [hunk 3]
```diff
@@ -0,0 +1,739 @@
...
+
+"The general problem of reaching Byzantine agreement on any value is reduced to the simple problem of reaching binary Byzantine agreement not eh nodes participating in each agreement protocol". [[5]]
```
Comment (CjS77) [inline src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md]: eh?

File: src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+
+Similar to Honeybadger BFT, this protocol is composed through the additions of interesting ideas presented in literature. 
+
+Like Hashgraph and Avalanche, a gossip protocol is used to allow efficient communication between nodes. [[1]]
+
+"The general problem of reaching Byzantine agreement on any value is reduced to the simpler problem of reaching binary Byzantine agreement on the nodes participating in each decision. This allows it to reuse the elegant binary Byzantine agreement protocol after adapting it to the gossip protocol". [[2]]
+
+Finally, the need for a trusted leader or a trusted setup phase implied is removed by porting the key ideas to an asynchronous setting [[3]]
+
+The network is set to *N of N* instances of the algorithm communicating via randomly synchronous connections.
```
Comment (CjS77) [inline src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md]: ???

File: src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md [hunk 5]
```diff
@@ -0,0 +1,739 @@
...
+
+## Democratic BFT
+
+This is a deterministic Byzantine consensus algorithm that relies on a new weak coordinator. This protocol is implemented in the Red Belly Block chain and is said to achieve 30 000 transactions/second on Amazon Cloud Trials [[36]] 
+
+The term weak coordinator is used to describe the ability of the algorithm to terminate in the presence of a faulty or slow coordinator unlike previous algorithms that do not have the ability to terminate. The fundamental idea here is to allow processes to complete asynchronous rounds as soon as they receive a threshold of messages, instead of having to wait for a message from a coordinator that may be slow. 
+
+The resulting algorithm assumes partial synchrony, is resilience optimal, time optimal and does not require signatures. 
+
+Moving away from the impossibility of solving consensus in asynchronous message systems, where processes can be faulty or *Byzantine*, the technic of randomization or additional synchrony is adopted.
```
Comment (CjS77) [inline src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md]: technic?

File: src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md [hunk 6]
```diff
@@ -0,0 +1,739 @@
...
+
+The term weak coordinator is used to describe the ability of the algorithm to terminate in the presence of a faulty or slow coordinator unlike previous algorithms that do not have the ability to terminate. The fundamental idea here is to allow processes to complete asynchronous rounds as soon as they receive a threshold of messages, instead of having to wait for a message from a coordinator that may be slow. 
+
+The resulting algorithm assumes partial synchrony, is resilience optimal, time optimal and does not require signatures. 
+
+Moving away from the impossibility of solving consensus in asynchronous message systems, where processes can be faulty or *Byzantine*, the technic of randomization or additional synchrony is adopted. 
+
+Randomized algorithms can used per-process "local" coins or a shared *common coin* to solve consensus probabilistically among *n* processes despite $t<n/3$ Byzantine processes. When based on local coins, the existing algorithms converge *O(n*<sup>2.5</sup>*)* expected time.
```
Comment (CjS77) [inline src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md]: used?

File: src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md [hunk 7]
```diff
@@ -0,0 +1,739 @@
...
+
+This protocol sees the use of a weak coordinator; a weak coordinator allows for the introduction of a new deterministic Byzantine consensus algorithm that is time optimal, resilience optimal and does not require the use of signatures. Unlike the classic, strong coordinator, the weak coordinator does not impost its value. It allows non-faulty processes to decide a value quickly, without the need of the coordinator, while helping the algorithm to terminate if non-faulty processes know that they proposed distinct values that might all be decided. In addition, the presence of a weak coordinator allows rounds to be executed optimistically without waiting for a specific message. This is unlike classic BFT algorithms that have to wait for a particular message from their coordinator and occasionally has to recover from a slow network or faulty coordinator.
```
Comment (CjS77) [inline src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md]: impost?

File: src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md [hunk 8]
```diff
@@ -0,0 +1,739 @@
...
+
+This protocol sees the use of a weak coordinator; a weak coordinator allows for the introduction of a new deterministic Byzantine consensus algorithm that is time optimal, resilience optimal and does not require the use of signatures. Unlike the classic, strong coordinator, the weak coordinator does not impost its value. It allows non-faulty processes to decide a value quickly, without the need of the coordinator, while helping the algorithm to terminate if non-faulty processes know that they proposed distinct values that might all be decided. In addition, the presence of a weak coordinator allows rounds to be executed optimistically without waiting for a specific message. This is unlike classic BFT algorithms that have to wait for a particular message from their coordinator and occasionally has to recover from a slow network or faulty coordinator. 
+
+With regards to the problem of a slow of Byzantine coordinator, the weak coordinator helps agreement by contributing a value while still allowing termination in a constant number of message delays and thus is unlike the classic coordinator or the eventual leader which cannot be implemented in BAMP<sub>n,t</sub>[*t<n/3*].
```
Comment (CjS77) [inline src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md]: what is BAMP?

File: src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md [hunk 9]
```diff
@@ -0,0 +1,749 @@
...
+- [Summary](#summary)
+- [References](#references)
+
+## Introduction 
+
+When considering the how Tari will potentially build its second layer, an analysis of the most promising Byzantine Consensus Mechanisms and their applications was sought. 
+
+Important to consider is the 'scalability trilemma'; a name referred to by Vitalik Buterin, which takes into account the potential trade-offs regarding decentralization, security and scalability. 
+
+**Decentralization** : a core principle on which majority of the systems are build, taking into account censorship-resistance and ensuring that everyone, without prejudice, is permitted to partake in the decentralized system.
```
Comment (hansieodendaal) [inline src/consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md]: Consistent use of <no space>:

**Comment by hansieodendaal** [comment]
Hi @Kevoulee , my review changes below. I will update this comment block as I make progress with the review.

2018-10-15

- [ ] With Introduction, ensure consistent use of 'some_text<no space>:<space> some text'

- [ ] Expand the TOC to include all headings

- [ ] Ensure all abbreviations are defined upon 1st use, example PoW, PBFT, FLP, DoS, DDoS, etc.

- [ ] Sentence logic problem "_Security: refers to the immutability of the ledger and takes into account, protects against the threat of 51% attacks, Sybil attacks and DDoS attacks etc._". Suggest "_Security: refers to the immutability of the ledger and takes into account threats of 51% attacks, Sybil attacks and DDoS attacks etc._"

- [ ] Be consistent with 'report' vs. 'paper'; "_This report investigates ..._" vs. "_This paper sees the ..._"

- [ ] Use of comma suggested: "_protocol assumptions; reference implementations_" vs. "_protocol assumptions, reference implementations_"

- [ ] Sentence not logical: "_There are a host own formal requirements ..._"

- [ ] It is not clear what is meant here: "_For systems with n processors, of which f are Byzantine, it ..._". Nodes have been introduced but processors not. Byzantine are also open to interpretation. Standardize on either nodes or processors throughout or state they can be used interchangeably.

- [ ] "number of involved **replicase** increases" -> "replicas"

- [ ] Give short explanation of "_liveness_" in "_However, the liveness properties of weakly synchronous protocols _"

- [ ] "_There **many** be periods of arbitrarily long days_" - > "_may_"

- [ ] Shortly define "_atomic broadcast_". Is it timeous like an atomic clock or at the same time like an atomic swap?

- [ ] Paragraph repeated:
       Protocols based on timing assumptions are unsuitable for decentralized, cryptocurrency settings, where network links can be unreliable, network speeds change rapidly, and network delays may even be adversarially induced. [6]
        The problem with both synchronous and partially synchronous assumptions is that "the protocols based on timing assumptions are unsuitable for decentralized, cryptocurrency settings, where network links can be unreliable, network speeds change rapidly, and network delays may even be adversarially induced."[6]

- [ ] Plural need here: "_While the vast majority of practical BFT **protocol** steer ..._"

- [ ] This sentence does not make sense, suggest change text in bold: "_While the vast majority of practical BFT **protocol** steer clear of this impossibility **result** by making timing assumptions, randomness (and, in particular, cryptography) **providers** an alternative route._" 

- [ ] Add caption for Figure 1 before referring to it: "_The gossip history can be represented as a directed graph, as in Figure 1._"

- [ ] Problem with link: "_`See [HashGraph](#HashGraph)`:_ ". Try "_`See [HashGraph](#hashgraph):`_"

- [ ] With Gossip, be consistent in use of italics for x, y, z, x1, etc.

- [ ] "they **havve** just learnt"

- [ ] The value system presented in the Summary must be discussed and shown in a separate section. For example, is 'Asymptotic Security' good, bad or just different? Some measurable parameters can be shown here for example rounds of communication between participating nodes. The scoring for each must be motivated if not really obvious. This can be done in a short paragraph for each protocol. It must be clear from the discussion which protocols rank highest for crypto-currency applications. How many registered nodes must part of a decision? If we have 10 nodes, is it still practical? What about 100 or 1000? Which of these have a practical track record? Which are only white papers? Group the value system in terms of decentralization, security and scalability. 

- [ ] The Summary must propose the most promising protocols to be pursued for the Tari protocol.

Thanks,
Hansie.

**Comment by CjS77** [comment]
Looks good. I still think you should split the report into 3-4 .md files though

**Comment by hansieodendaal** [comment]
Hi @CjS77 and @Kevoulee , I misunderstood this comment, "_The report is very long, so I suggest breaking it up into 3-4 sections_", otherwise I would have spoken up earlier. Now I see what was meant is "_I still think you should split the report into 3-4 .md files though_". 

Let us discuss face to face, as this has a big impact on this report as well as other reports we have been writing. We have different value systems at play here, and we should converge sooner rather than later.

Let me illustrate: This work has not been designed as a collection of reports, and my typical next comment before I approve the changes would be to combine all the reports into one report again or to make each file into a complete stand-alone report.

Please give a thumbs up for us to have a meeting.

Thanks.

**Comment by CjS77** [comment]
@Kevoulee you need to get the tests to pass.

Most likely it's ``` sections that need to change to 
<pre>
```text
blah blah blah
```
</pre>

Looks carefully at the [logs](https://travis-ci.org/tari-labs/tari-university/builds/455851678?utm_source=github_status&utm_medium=notification) to find the errors

**Comment by hansieodendaal** [comment]
@CjS77 and @Kevoulee , the problem is with 'protocols/grin-protocol-overview/MainReport.md' merged yesterday, not with this PR or branch. Will fix in another branch/PR.

**Comment by CjS77** [comment]
Wow, that shouldn't have been merged in then.

**Comment by CjS77** [comment]
Failing tests usually mean the site / code is broken, and we shouldn't have broken code in master.

**Comment by hansieodendaal** [comment]
My bad, did not realize , found the error

**Pull request merged**